### PR TITLE
Bank tags: tag charge/degradable items

### PR DESCRIPTION
**Pull request by raiyni**
Holding shift adds the ability to edit variation tags on an item.

peek 2018-10-30 09-04

Closes https://github.com/runelite/runelite/issues/3767

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java [hunk 1]
```diff
@@ -58,6 +60,18 @@
 @Slf4j
 public class ItemManager
 {
+	private static final List<Pattern> normalizePatterns = new ArrayList<Pattern>()
+	{{
+		// cakes, pizzas, pies
+		add(Pattern.compile("^(?:[0-9]/[0-9]|half an?) ([^(0-9]+)|$"));
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: [0-9]/[0-9] could probably be [1-9]/[1-9]? There's no items that are "0/3" for instance, and I doubt there will ever be 0 in the denominator either.

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java [hunk 2]
```diff
@@ -58,6 +60,18 @@
 @Slf4j
 public class ItemManager
 {
+	private static final List<Pattern> normalizePatterns = new ArrayList<Pattern>()
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: Is there any way we can run a test for this? Seems like we'd want to make sure this doesn't break with future updates.

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java [hunk 3]
```diff
@@ -307,4 +332,28 @@ public MouseWheelEvent mouseWheelMoved(MouseWheelEvent event)
 		clientThread.invokeLater(() -> tabInterface.handleWheel(event));
 		return event;
 	}
+
+	@Override
+	public void keyTyped(KeyEvent e)
+	{
+
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java]: Remove this empty line I guess

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java (hunk 2 above)
Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: also, UPPER_CASE

**Comment by raiyni** [comment]
I'm writing tests

File: runelite-client/src/main/java/net/runelite/client/game/ItemManager.java [hunk 4]
```diff
@@ -58,6 +60,18 @@
 @Slf4j
 public class ItemManager
 {
+	private static final List<Pattern> UNCHARGE_PATTERNS = new ArrayList<Pattern>()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/game/ItemManager.java]: ImmutableList.builder

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java [hunk 5]
```diff
@@ -108,6 +112,11 @@
 	@Inject
 	private TabInterface tabInterface;
 
+	@Inject
+	private KeyManager keyManager;
+
+	private boolean shiftPressed = false;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/BankTagsPlugin.java]: You should reset shiftPressed on focus lost and shutdown

File: runelite-client/src/main/java/net/runelite/client/plugins/banktags/TagManager.java [hunk 6]
```diff
@@ -85,20 +131,48 @@ public void addTag(int itemId, String tag)
...
+		return tags.stream().anyMatch(tag -> tag.contains(Text.standardize(search)));
 	}
 
 	public void removeTag(String tag)
 	{
 		final String prefix = CONFIG_GROUP + "." + ITEM_KEY_PREFIX;
 		configManager.getConfigurationKeys(prefix).forEach(item -> removeTag(Integer.parseInt(item.replace(prefix, "")), tag));
+
+		final String chargePrefix = CONFIG_GROUP + "." + VARIATION_KEY_PREFIX;
+		configManager.getConfigurationKeys(prefix).forEach(item -> removeVariationTag(Integer.parseInt(item.replace(chargePrefix, "")), tag));
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/banktags/TagManager.java]: you can get the configuration keys only once and reuse

Comment (deathbeam) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/TagManager.java]: or, do getConfigurationKeys(chargePrefix) as that looks like correct behaviour here

Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/banktags/TagManager.java]: yeah it's a typo

**Comment by raiyni** [comment]
now with passing tests

**Comment by raiyni** [comment]
Rebased to use an enum

**Comment by Nightfirecat** [comment]
So the drag-onto-tab behavior for item variations is only changed when holding shift, right? I don't know that I'd often want to tag variations this way so I'd be against the default behavior changing.

**Comment by raiyni** [comment]
Yes, only when holding shift

**Comment by deathbeam** [comment]
I am afraid this will not work with import/export very well and that I cant really make it work reliably without the exported string being some insanity and the website handling code as well

**Comment by raiyni** [comment]
Updated to use current config tag key. Should be compatible with import/export. Up to the website to display it right.

**Comment by deathbeam** [comment]
Updated the PR, as I said in Discord:

```
Updated the variation tag PR, it should be a bit cleaner now. Also, made remove-tag remove both variation tags and normal tags, because otherwise it will not work (e.g you have item tagged only with variation tag but you try to remove it as regular tag, nothing happens)
```

**Comment by deathbeam** [comment]
Updated the PR (unified edit-tags and edit-variation-tags) it works like this now:

peek 2018-10-30 09-04

**Pull request merged**